src/ripple/app/tx/impl/Remit.cpp

```diff
@@ -0,0 +1,545 @@
+                if (!sb.dirRemove(
+                        keylet::ownerDir(srcAccID), page, kl.key, true))
+                {
+                    JLOG(j.fatal())
+                        << "Could not remove URIToken from owner directory";
+                    return tefBAD_LEDGER;
+                }
+
+                adjustOwnerCount(sb, sleSrcAcc, -1, j);
+            }
+
+            // add to dest dir
+            {
+                auto const page = sb.dirInsert(
+                    keylet::ownerDir(dstAccID), kl, describeOwnerDir(dstAccID));
+
+                JLOG(j_.trace()) << "Adding URIToken to owner directory "
+                                 << to_string(kl.key) << ": "
+                                 << (page ? "success" : "failure");
+
+                if (!page)
+                    return tecDIR_FULL;
+
+                sleU->setFieldU64(sfOwnerNode, *page);
+
+                adjustOwnerCount(sb, sleDstAcc, 1, j);
+            }
+
+            // change the owner
+            sleU->setAccountID(sfOwner, dstAccID);
+
+            sb.update(sleU);
+        }
+    }
+
+    // iterate trustlines
+    if (ctx_.tx.isFieldPresent(sfAmounts))
+    {
+        // process trustline remits
+        STArray const& sEntries(ctx_.tx.getFieldArray(sfAmounts));
+        for (STObject const& sEntry : sEntries)
+        {
+            STAmount const amount = sEntry.getFieldAmount(sfAmount);
+            if (isXRP(amount))
+            {
+                // since we have to pay for all the created objects including
+                // possibly the account itself this is paid right at the end,
+                // and only if there is balance enough to cover.
+                nativeRemit += amount.xrp();
+                continue;
+            }
+
+            AccountID const issuerAccID = amount.getIssuer();
+
+            // check permissions
+            if (issuerAccID == srcAccID || issuerAccID == dstAccID)
+            {
+                // no permission check needed when the issuer sends out or a
+                // subscriber sends back RH TODO: move this condition into
+                // trustTransferAllowed, guarded by an amendment
+            }
+            else if (TER canXfer = trustTransferAllowed(
+                         sb,
+                         std::vector<AccountID>{srcAccID, dstAccID},
+                         amount.issue(),
+                         j);
+                     canXfer != tesSUCCESS)
+                return canXfer;
+
+            // compute the amount the source will need to send
+            // in remit the sender pays all transfer fees, so that
+            // the destination can always be assured they got the exact amount
+            // specified. therefore we need to compute the amount + transfer fee
+            auto const srcAmt =
+                issuerAccID != srcAccID && issuerAccID != dstAccID
+                ? multiply(amount, transferRate(sb, issuerAccID))
```

Comment:
I don't think this transfer rate is tested anywhere? given history of getting this wrong please add test cases for 0 xfer rate and non-zero as well

Reply:
I tested it the same way we did in URIToken after the fix. https://github.com/Xahau/xahaud/blob/4e8c0b4388eaac006e71dd7dd9bca4a0c73e5d5c/src/test/app/Remit_test.cpp#L2017

Will add test for 0 and non 0.